File: leetcode/532.k-diff-pairs-in-an-array.py
class Solution(object):
    def findPairs(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        
        if nums is None:
            return 0

        nums.sort()
        start = 0
        ans = 0

        while start < len(nums):
            end = start + 1

            while(end < len(nums) and nums[end] - nums[start] < k):
                end += 1

            if (end < len(nums) and nums[end] - nums[start] == k):
                ans += 1

                #remove duplicate
                while(start + 1 < len(nums) and nums[start + 1] == nums[start]):
                    start += 1

            start += 1

        return ans

File: leetcode/532.k-diff-pairs-in-an-array.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int ans = 0;
        for (int i = 0; i < nums.size(); ++i)
        {
            int j = i + 1;
            
            while(j < nums.size() && nums[j] - nums[i] < k)
            {
                j++;
            }
            
            if (j < nums.size() && nums[j] - nums[i] == k)
            {
                ans += 1;
                
                // remove duplicate
                while(i + 1 < nums.size() && nums[i + 1] == nums[i])
                {
                    i++;
                }
            }
        }
        
        return ans;
    }
};